### PR TITLE
[ML][Data Frames] Fix null aggregation handling in indexer

### LINKED ISSUE #44906: [ML-DataFrame] NPE if no source index available / revisit empty set of source indexes
Originally at lease 1 source index was required, in case of a pattern new index could appear but it wasn't allowed to remove them. To support ILM however we decided to relax this constraint and allow the deletion of indexes, see https://github.com/elastic/elasticsearch/pull/44344.

As indexes could get deleted while indexing is running it could happen that searches fail if the index is closed at the moment we try to access it, that's why we had to switch to lenient source index expansion: https://github.com/elastic/elasticsearch/pull/44633

This had a bad side effect: sources indexes can run empty which breaks assumptions made in various places of the indexer, e.g. leading to a NPE:

```
[2019-07-25T06:49:38,936][WARN ][o.e.x.d.t.DataFrameTransformTask] [ml-qa-ubuntu-5] Data frame transform [dft_failed_missing_source_index_1564037378655] encountered an exception: 
java.lang.NullPointerException: null
    at org.elasticsearch.xpack.dataframe.transforms.DataFrameIndexer.doProcess(DataFrameIndexer.java:170) ~[?:?]
    at org.elasticsearch.xpack.core.indexing.AsyncTwoPhaseIndexer.onSearchResponse(AsyncTwoPhaseIndexer.java:337) ~[?:?]
    at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:62) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
    at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
    at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:68) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
```

(the NPE happens as we try to access the non-existing `aggs` object in the response)

## Solution

 - ~if a concrete index is involved and the set of source indexes gets empty the transform should fail~
 - ~if the source is a pattern~ we allow source getting empty

=> always be lenient (whether or not source is a pattern)
=> require a non-empty source(index, not document-wise!) on `_preview` and `PUT`

## Dev Notes

 - check for empty source set / results:
   - at validation time if validation is not deferred do not allow an empty set, independent of pattern usage
   - at preview, independent of pattern usage
 - expect empty/null results, code needs to be hardened for that

## Discuss

=> a strict mode will be provided once we see a demand for it

Outdated:

The deletion of source indexes seem only relevant if source indexes are managed by ILM and for continuous data frames:

- should batch data frames be strict like before?
- even if continuous is used you might not want to "loose" data without notice, should we have a parameter for strictness (while lenient can be the default)?
  - or use `index` for non-ILM and `index_patterns` for ILM:

non-ILM:
```
"source": {
    "index": "kibana_sample_data_ecommerce",
    "query": {}
...
```

ILM: 
```
"source": {
    "index_patterns": ["events-*"],
    "query": {}
...
```

### PR DESCRIPTION
**Pull request by benwtrent**
Since we are more lenient now when expanding index patterns, it could be that there are no open concrete indices contained within the pattern. When this is the case `SearchResponse#getAggregations()` returns `null`. 

This adds a `null` check for that method, audits that possibly all the underlying indices are closed/deleted, and returns that the current checkpoint is done. 

closes https://github.com/elastic/elasticsearch/issues/44906

### CODE REVIEW

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java [hunk 1]
```diff
@@ -167,7 +169,20 @@ protected void onFinish(ActionListener<Void> listener) {
...
+        // Treat this as a "we reached the end".
+        // This should only happen when all underlying indices have gone away. Consequently, there is no more data to read.
+        if (aggregations == null) {
+            logger.info("[" + getJobId() + "] unexpected null aggregations in search response. " +
+                "Concrete indices may have disappeared or have been closed");
+            auditor.info(getJobId(),
+                "Data frame transform unexpectedly stopped seeing data. " +
+                    "Please verify that these indices exist and are open [" +
+                    Strings.arrayToCommaDelimitedString(getConfig().getSource().getIndex()) +
+                    "].");
```
Comment (benwtrent) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java]: I waffled a bit on how to word this. From my understanding, `aggregations` should only ever be null here iff the underlying indices are all closed/deleted

**Comment by benwtrent** [comment]
run elasticsearch-ci/packaging-sample

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java (hunk 1 above)
Comment (hendrikmuhs) [reply x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java]: yes, it's only `null` if source became empty.

With the given wording, it sounds more like a warning. So I would either change the log/audit level or the wording.

**Review by hendrikmuhs — COMMENTED** [review]
What about other places in the code? My gut feeling about #44906: it's more than this, see the dev notes about `_preview` and deferred validation.

**Comment by benwtrent** [comment]
@hendrikmuhs we already validate for concrete index existence on `_start` and on `_preview`.
Preview: https://github.com/elastic/elasticsearch/blob/204ed8509a4687f8e5067f825a14fe822aad46d0/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportPreviewDataFrameTransformAction.java#L94..L103

Start: https://github.com/elastic/elasticsearch/blob/204ed8509a4687f8e5067f825a14fe822aad46d0/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStartDataFrameTransformAction.java#L184..L185

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportPreviewDataFrameTransformAction.java [hunk 2]
```diff
@@ -155,7 +156,15 @@ private void getPreview(Pivot pivot,
                     ActionListener.wrap(
                         r -> {
                             try {
-                                final CompositeAggregation agg = r.getAggregations().get(COMPOSITE_AGGREGATION_NAME);
+                                final Aggregations aggregations = r.getAggregations();
+                                if (aggregations == null) {
+                                    listener.onFailure(
+                                        new ElasticsearchStatusException("Underlying source indices have gone away",
```
Comment (droberts195) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportPreviewDataFrameTransformAction.java]: How about `Source indices have been deleted or closed`?

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java [hunk 3]
```diff
@@ -167,7 +169,20 @@ protected void onFinish(ActionListener<Void> listener) {
 
     @Override
     protected IterationResult<DataFrameIndexerPosition> doProcess(SearchResponse searchResponse) {
-        final CompositeAggregation agg = searchResponse.getAggregations().get(COMPOSITE_AGGREGATION_NAME);
+        final Aggregations aggregations = searchResponse.getAggregations();
+        // Treat this as a "we reached the end".
+        // This should only happen when all underlying indices have gone away. Consequently, there is no more data to read.
+        if (aggregations == null) {
+            logger.info("[" + getJobId() + "] unexpected null aggregations in search response. " +
+                "Concrete indices may have disappeared or have been closed");
```
Comment (droberts195) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java]: Again, `Source indices have been deleted or closed` for the second sentence

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java [hunk 4]
```diff
@@ -167,7 +169,20 @@ protected void onFinish(ActionListener<Void> listener) {
...
+            auditor.info(getJobId(),
+                "Concrete source indices seem to have gone away. " +
```
Comment (droberts195) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java]: `Source indices have been deleted or closed` for the first sentence

**Review by droberts195 — COMMENTED** [review]
I don't think we need the words "Concrete" or "Underlying" in the messages, but apart from that I'm happy with this change

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by hendrikmuhs — APPROVED** [review]
LGTM

I am not sure about the logging/audit w.r.t. a PR I am working on, but this can be revisited as a follow up.

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by benwtrent** [comment]
run elasticsearch-ci/2

**Pull request merged**